tool/site/client/src/components/messageExpanders.tsx

```diff
@@ -19,45 +20,90 @@ const MessageExpanders: React.FC<MessageExpandersProps> = ({
   hl7V3Response
 }) => (
   <Details.ExpanderGroup>
-    <MessageExpander
+    <JsonMessageExpander
       name="Request (FHIR)"
-      message={JSON.stringify(fhirRequest, null, 2)}
-      mimeType="application/json"
+      message={fhirRequest}
     />
-    {hl7V3Request && <MessageExpander
+    {hl7V3Request && <XmlMessageExpander
       name="Request (HL7 V3)"
       message={hl7V3Request}
-      mimeType="text/xml"
     />}
-    <MessageExpander
+    <JsonMessageExpander
       name="Response (FHIR)"
-      message={JSON.stringify(fhirResponse, null, 2)}
-      mimeType="application/json"
+      message={fhirResponse}
     />
-    {hl7V3Response && <MessageExpander
+    {hl7V3Response && <XmlMessageExpander
       name="Response (HL7 V3)"
       message={hl7V3Response}
-      mimeType="text/xml"
     />}
   </Details.ExpanderGroup>
 )
 
-interface MessageExpanderProps {
+const StyledButton = styled(Button)`
+  margin-bottom: 0;
+`
+
+interface JsonMessageExpanderProps {
+  name: string
+  message: FhirResource
+}
+
+interface XmlMessageExpanderProps {
   name: string
   message: string
-  mimeType: string
 }
 
-const StyledButton = styled(Button)`
-  margin-bottom: 0;
-`
+export const JsonMessageExpander: React.FC<JsonMessageExpanderProps> = ({
+  name,
+  message
+}) => {
+  const downloadHref = `data:application/json;charset=utf-8,${encodeURIComponent(JSON.stringify(message, null, 2))}`
+  return (
+    <Details expander>
+      <Details.Summary>{name}</Details.Summary>
+      <Details.Text>
+        <ButtonList>
+          <StyledButton onClick={() => navigator.clipboard.writeText(JSON.stringify(message, null, 2))}>Copy</StyledButton>
+          <StyledButton download="message" href={downloadHref}>Download</StyledButton>
+        </ButtonList>
+        <ReactJson
+          collapseStringsAfterLength={50}
+          displayDataTypes={false}
+          displayObjectSize={false}
+          name={false}
+          src={message}
+          style={{marginTop: "10px"}}
+          theme={
+            {
```

Comment:
would u mind making this a separate const outside of this scope just for readability and stuff

Reply:
moved